### PR TITLE
Contracts and harnesses for `ptr::Unique`

### PR DESCRIPTION
**Pull request by tautschnig**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/core/src/ptr/unique.rs [hunk 1]
```diff
@@ -84,13 +88,16 @@ impl<T: ?Sized> Unique<T> {
...
     #[inline]
+    #[requires(!ptr.is_null())]
     pub const unsafe fn new_unchecked(ptr: *mut T) -> Self {
         // SAFETY: the caller must guarantee that `ptr` is non-null.
         unsafe { Unique { pointer: NonNull::new_unchecked(ptr), _marker: PhantomData } }
     }
 
     /// Creates a new `Unique` if `ptr` is non-null.
     #[inline]
+    #[ensures(|result| result.is_none() == result.unwrap().as_ptr().is_null())]
```
Comment (feliperodri) [inline library/core/src/ptr/unique.rs:99]: Wouldn't this panic?

File: library/core/src/ptr/unique.rs [hunk 2]
```diff
@@ -201,3 +209,92 @@ impl<T: ?Sized> From<NonNull<T>> for Unique<T> {
...
+
+#[cfg(kani)]
+#[unstable(feature="kani", issue="none")]
+mod verify {
+    use super::*;
+
+    // pub const unsafe fn new_unchecked(ptr: *mut T) -> Self
+    #[kani::proof_for_contract(Unique::new_unchecked)]
+    pub fn check_new_unchecked() {
+        let mut x : i32 = 42;
```
Comment (feliperodri) [inline library/core/src/ptr/unique.rs:221]:
```suggestion
        let mut x : i32 = kani::any();
```

Comment (feliperodri) [reply library/core/src/ptr/unique.rs:221]: why not use `kani::any()` in all these harnesses?

File: library/core/src/ptr/unique.rs [hunk 3]
```diff
@@ -201,3 +209,92 @@ impl<T: ?Sized> From<NonNull<T>> for Unique<T> {
...
+    }
+
+    // pub const fn new(ptr: *mut T) -> Option<Self>
+    #[kani::proof_for_contract(Unique::new)]
+    pub fn check_new() {
+        let mut x : i32 = 42;
+        let xptr = &mut x;
+
+        if let Some(unique) = Unique::new(xptr as *mut i32) {
+            assert_eq!(unique.as_ptr(), xptr as *mut i32);
```
Comment (feliperodri) [inline library/core/src/ptr/unique.rs:237]: Why these assertions are not `ensures` clauses?

**Review by feliperodri — CHANGES_REQUESTED** [review]

File: library/core/src/ptr/unique.rs (hunk 2 above)
Comment (tautschnig) [reply library/core/src/ptr/unique.rs:221]: Oh, sure, sorry. Done.

File: library/core/src/ptr/unique.rs (hunk 1 above)
Comment (tautschnig) [reply library/core/src/ptr/unique.rs:99]: Fixed by using `ptr` on the right-hand side.

File: library/core/src/ptr/unique.rs (hunk 3 above)
Comment (tautschnig) [reply library/core/src/ptr/unique.rs:237]: Done where feasible.

**Review by feliperodri — APPROVED** [review]
LGTM.

**Pull request merged**